### PR TITLE
Make Tuple use a dynamic default when necessary

### LINKED ISSUE #1234: Invalid error messages from nested collections
The error messages produced by operations on nested collections are incorrect.

After:
```python
>>> class A(HasTraits):
...     foo = List(List(Int))
... 
>>> a = A(foo=[[1, 2], [3, 4]])
```

We get this exception from a bad operation on an inner list:

```python
>>> a.foo[0].append(3.7)

Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_types.py", line 218, in validate
    return _validate_int(value)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_types.py", line 168, in _validate_int
    return int(operator.index(value))
TypeError: 'float' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_list_object.py", line 723, in append
    super().append(object)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_list_object.py", line 361, in append
    super().append(self.item_validator(object))
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_list_object.py", line 862, in _item_validator
    return trait_validator(object, self.name, value)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_types.py", line 220, in validate
    self.error(object, name, value)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/base_trait_handler.py", line 74, in error
    raise TraitError(
traits.trait_errors.TraitError: Each element of the 'foo' trait of an A instance must be an integer, but a value of 3.7 <class 'float'> was specified.
```

The statement that "Each element of the 'foo' trait of an A instance must be an integer" is incorrect: each element of the 'foo' trait of an A instance should be a list of integers, not an integer.

### LINKED ISSUE #1520: Tuple has bad defaults for some inner trait types (for example, List)
The `Tuple` trait default value is currently always static, regardless of the inner trait types. This causes issues with a trait definition of the form `Tuple(Int, List(Str))`:

- no validation is performed on the inner list
- if the trait type is re-used, the list is shared

Example of the failure of validation:

```python
>>> from traits.api import *
>>> class A(HasTraits):
...     foo = Tuple(Int, List(Int))
... 
>>> a = A()
>>> a.foo[0]
0
>>> a.foo[1].append("not a number")  # Should fail, but doesn't
>>> a.foo
(0, ['not a number'])
```

Continuing the above example:
```
>>> b = A()
>>> b.foo
(0, ['not a number'])  # expected (0, [])
```

The `Union` trait type appears to have similar issues.

### PR DESCRIPTION
**Pull request by mdickinson**
This PR fixes the `Tuple` trait type so that it only uses a constant default value when all of its component traits can use a constant default value. This includes simple cases like `Tuple(Int(), Str())` as well as nested cases like `Tuple(Int(), Tuple(Int(), Str()))`.

Fixes #1520.

### CODE REVIEW

File: traits/trait_types.py [hunk 1]
```diff
@@ -2283,16 +2283,40 @@ def __init__(self, *types, **metadata):
...
+                self.default_value_type = DefaultValue.callable
+                default_value = self._get_default_value
 
         super().__init__(default_value, **metadata)
 
+    def _get_default_value(self, object):
+        # Dynamic default, used when at least one of the child traits requires
+        # a dynamic default.
+        return tuple(
+            inner_trait.default_value_for(object, "<inner_trait>")
```
Comment (mdickinson) [inline traits/trait_types.py:2316]: This part's annoying: we have to invent a name because `default_value_for` expects one, but the `DefaultValue.callable` default type doesn't pass any name along. So I chose something that should at least be non-ambiguous and can easily be searched for in the Traits source.

I couldn't find any way to actually get the name to show up in validation error messages. For example with `List` child traits, at the point where this dynamic default is used, it's then validated and at the validation step the `TraitList` objects are recreated with name matching the name of the actual trait; those recreated `TraitList` objects are what are actually used in the trait value.

File: traits/trait_types.py [hunk 2]
```diff
@@ -2283,16 +2283,40 @@ def __init__(self, *types, **metadata):
             if len(types) == 0:
                 types = [Trait(element) for element in default_value]
 
-        self.types = tuple([trait_from(type) for type in types])
+        self.types = tuple(trait_from(type) for type in types)
```
Comment (mdickinson) [inline traits/trait_types.py:2286]: Gratuitous drive-by style cleanup.

File: traits/trait_types.py [hunk 3]
```diff
@@ -2283,16 +2283,40 @@ def __init__(self, *types, **metadata):
...
         if default_value is None:
-            default_value = tuple(
-                [type.default_value()[1] for type in self.types]
+            # Optimisation: if all child traits have a constant default value,
+            # we can use a constant default value too. Otherwise the default
+            # needs to be computed dynamically.
+            child_defaults = []
+            child_default_types = []
+            for child_trait in self.types:
+                default_type, default_value = child_trait.default_value()
```
Comment (mdickinson) [inline traits/trait_types.py:2296]: Nitpick: nasty overloading of the `default_value` name here. I'll fix.

File: traits/trait_types.py [hunk 4]
```diff
@@ -2283,16 +2283,41 @@ def __init__(self, *types, **metadata):
...
                 types = [Trait(element) for element in default_value]
 
-        self.types = tuple([trait_from(type) for type in types])
+        self.types = tuple(trait_from(type) for type in types)
         self.init_fast_validate(ValidateTrait.tuple, self.types)
 
         if default_value is None:
-            default_value = tuple(
-                [type.default_value()[1] for type in self.types]
+            # Optimisation: if all child traits have a constant default value,
```
Comment (mdickinson) [inline traits/trait_types.py:2290]: Hmm. This isn't actually purely optimisation: we don't validate defaults in the static case, so in current Traits we can have for example:

```
>>> from traits.api import *
>>> class A(HasTraits):
...     foo = Tuple(Int("some string"), Str(5))
... 
>>> a = A()
>>> a.foo
('some string', 5)
```

The "optimisation" here ensures that this behaviour (or arguably misbehaviour) persists.

File: traits/trait_types.py [hunk 5]
```diff
@@ -2283,16 +2283,41 @@ def __init__(self, *types, **metadata):
...
+            child_default_types = []
+            for child_trait in self.types:
+                child_default_type, child_default = child_trait.default_value()
+
+                child_default_types.append(child_default_type)
+                child_defaults.append(child_default)
+
+            constant_default = all(
+                dvt == DefaultValue.constant for dvt in child_default_types
             )
```
Comment (corranwebster) [inline traits/trait_types.py:2303]: Short-circuit the loop if you hit a non-constant default
```suggestion
            child_defaults = []
            constant_default = True
            for child_trait in self.types:
                child_default_type, child_default = child_trait.default_value()
                if child_default_type != DefaultValue.constant:
                    constant_default = False
                    break
                child_defaults.append(child_default)
```
or something like this?

**Review by corranwebster — APPROVED** [review]
Looks good.  One suggestion, but I don't know if it is an improvement or not.

File: traits/trait_types.py (hunk 5 above)
Comment (mdickinson) [reply traits/trait_types.py:2303]: I actually already did this, and then changed to the current implementation on the basis that it was clearer and that the loop seemed unlikely to be a bottleneck.

Comment (mdickinson) [reply traits/trait_types.py:2303]: (Off-topic, but turning a list of pairs into a pair of lists seems unnecessarily awkward in Python - it feels as though it ought to be a one-liner.)
